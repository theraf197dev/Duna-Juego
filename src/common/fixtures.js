import trialImage from './media/trial_mode.jpg';
import seekerImage from './media/seeker_mode.jpg';

const controls = [
    {
        id: 1,
        name: 'Up',
        vector: 'x',
        incr: -1,
    },
    {
        id: 2,
        name: 'Down',
        vector: 'x',
        incr: 1,
    },
    {
        id: 3,
        name: 'Left',
        vector: 'y',
        incr: -1,
    },
    {
        id: 4,
        name: 'Right',
        vector: 'y',
        incr: 1,
    },
];

const games = {
    items: [
        {
            name: 'TRIAL',
            url: trialImage,
        },
        {
            name: 'SEEKER',
            url: seekerImage,
        },
    ],
};

const difficulty = 1;
const mode = 'TRIAL';
const size = 9;

export {
    controls,
    difficulty,
    games,
    mode,
    size,
};
