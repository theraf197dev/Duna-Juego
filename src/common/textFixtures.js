export const texts = {
    difficultyText: 'Difficulty',
    gameDescription: 'Atrapa la pizzita',
    gameName: 'DUNA JUEGO',
    playButtonText: 'PLAY',
    retryButtonText: 'RETRY?',
};
