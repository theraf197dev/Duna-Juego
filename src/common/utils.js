import dunaImage from './media/duna.jpg';
import pizzaImage from './media/pizza.png';
import blankImage from './media/blank.png';
import obstacleImage from './media/obstacle.jpg';
import seekerImage from './media/seeker.jpg';

const CellType = {
    blank: 'blank',
    duna: 'duna',
    pizza: 'pizza',
    obstacle: 'obstacle',
    seeker: 'seeker',
};

const GameMode = {
    seeker: 'SEEKER',
    trial: 'TRIAL',
};

const GameStatus = {
    defeat: 'defeat',
    playing: 'playing',
    victory: 'victory',
};

const ImageCarousel = (name) => {
    let image;
    switch (name) {
        case CellType.duna:
            image = dunaImage;
            break;
        case CellType.pizza:
            image = pizzaImage;
            break;
        case CellType.blank:
            image = blankImage;
            break;
        case CellType.obstacle:
            image = obstacleImage;
            break;
        case CellType.seeker:
            image = seekerImage;
            break;
        default:
            image = blankImage;
    }
    return image;
};

const canMove = (matrix, currentPos, vector, incr, size) => {
    return !isOutOfBounds(currentPos, vector, incr, size) && !isObstacle(matrix, currentPos, vector, incr, size);
};

const isObstacle = (matrix, currentPos, vector, incr) => {
    const nextPos = {
        ...currentPos,
        [vector]: currentPos[vector] + incr,
    };
    return matrix[nextPos.x][nextPos.y] === CellType.obstacle;
};

const isOutOfBounds = (currentPos, vector, incr, size) => {
    return (currentPos[vector] + incr) < 0 || (currentPos[vector] + incr) >= size;
};

export {
    canMove,
    CellType,
    GameMode,
    GameStatus,
    ImageCarousel,
};
