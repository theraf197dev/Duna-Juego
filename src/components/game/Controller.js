import { Component } from "react";
import {
    CellType,
    GameMode,
    GameStatus,
} from '../../common/utils';
import View from './View.jsx';

export default class Controller extends Component {
    componentDidUpdate(prevProps) {
        if(prevProps.currentPos !== this.props.currentPos) {
            this.props.setMatrix(CellType.duna);
        }

        if(prevProps.seekerCurrentPos !== this.props.seekerCurrentPos) {
            this.props.setMatrix(CellType.seeker, this.props.seekerCurrentPos);
        }

        if(this.props.gameStatus !== GameStatus.playing && this.props.mode === GameMode.seeker) {
            this.props.stopInterval();
        }

        if(this.props.pizzas <= 0 && prevProps.pizzas !== this.props.pizzas) {
            this.props.endGame(GameStatus.victory);
        }
    }

    render() {
        return (
            <View
                {...this.props}
                endGame={this.props.endGame}
                setCurrentPos={this.props.setCurrentPos}
            />
        )
    }
};

Controller.displayName = 'ControllerGame';
