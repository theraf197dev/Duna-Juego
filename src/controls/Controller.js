import { Component } from "react";
import View from './View.jsx';
import {
    CellType,
} from '../common/utils';

export default class Controller extends Component {
    constructor(props) {
        super(props);
        this.bindActions();
    };

    bindActions() {
        this.handleChangePosition = this.handleChangePosition.bind(this);
        this.canMove = this.canMove.bind(this);
    }

    canMove(vector, incr) {
        return !this.isOutOfBounds(vector, incr) && !this.isObstacle(vector, incr);
    }

    isObstacle(vector, incr) {
        const nextPos = {
            ...this.props.currentPos,
            [vector]: this.props.currentPos[vector] + incr,
        };
        return this.props.matrix[nextPos.x][nextPos.y] === CellType.obstacle;
    }

    isOutOfBounds(vector, incr) {
        return (this.props.currentPos[vector] + incr) < 0 || (this.props.currentPos[vector] + incr) >= this.props.size;
    }

    handleChangePosition(vector, incr) {
        if(this.canMove(vector, incr)) {
            this.props.setCurrentPos(vector, incr);
            this.props.setLastInput({vector, incr});
        }
    }

    render() {
        return (
            <View
                {...this.props}
                handleChangePosition={this.handleChangePosition}
                canMove={this.canMove}
            />
        )
    }
}

