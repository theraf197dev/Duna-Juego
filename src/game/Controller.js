import { Component } from "react";
import {
    CellType,
} from '../common/utils';
import View from './View.jsx';

export default class Controller extends Component {
    componentDidUpdate(prevProps) {
        if(prevProps.currentPos !== this.props.currentPos){
            this.props.setMatrix(CellType.duna);
        }
    }

    render() {
        return (
            <View
                {...this.props}
                setCurrentPos={this.props.setCurrentPos}
            />
        )
    }
}

