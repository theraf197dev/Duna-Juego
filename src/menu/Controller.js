import { Component } from "react";
import View from './View.jsx';

export default class Controller extends Component {
    constructor(props) {
        super(props);
        this.bindActions();
    };

    bindActions() {
        this.initGame = this.initGame.bind(this);
    }

    initGame() {
        this.props.startGame(this.props.difficulty, this.props.mode);
    }

    render() {
        return (
            <View
                {...this.props}
                initGame={this.initGame}
            />
        )
    }
}

Controller.displayName = 'ControllerMenu';
