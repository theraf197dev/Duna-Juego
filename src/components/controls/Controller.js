import { Component } from "react";
import PropTypes from 'prop-types';
import View from './View.jsx';
import barkSFX from '../../common/media/bark.mp3';
import { GameStatus } from "../../common/utils.js";

export default class Controller extends Component {
    constructor(props) {
        super(props);
        this.bindActions();
        this.audio = new Audio(barkSFX);
    };

    bindActions() {
        this.handleChangePosition = this.handleChangePosition.bind(this);
        this.handleBark = this.handleBark.bind(this);
    };

    handleChangePosition(vector, incr) {
        this.props.setCurrentPos(vector, incr);
    };

    handleBark() {
        this.audio.play();
    };

    render() {
        return (
            <View
                {...this.props}
                handleChangePosition={this.handleChangePosition}
                handleBark={this.handleBark}
            />
        )
    }
}

Controller.propTypes = {
    controls: PropTypes.array,
    currentPos: PropTypes.object,
    gameStatus: PropTypes.string,
    matrix: PropTypes.array,
    setCurrentPos: PropTypes.func,
    size: PropTypes.number,
};

Controller.displayName= 'ControllerControls';
