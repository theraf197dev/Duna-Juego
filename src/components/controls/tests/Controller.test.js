import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import fixtures from '../../../common/fixtures';
import { GameStatus } from "../../../common/utils";
import Controller from "../Controller";

describe('Controller tests', () => {
  let props = {
    ...fixtures,
    currentPos: {
      x: 3,
      y: 4,
    },
    gameStatus: GameStatus.playing,
  };

  it('render component', () => {
      const { container } = render(<Controller {...props} />);

      expect(container).toMatchSnapshot();
  });

  it('handleChangePosition has been called when click on button', async () => {
    const testProps = {
      ...props,
      setCurrentPos: jest.fn(),
    };

      render(<Controller {...testProps} />);

      const upButton = await screen.findByTestId('fn-controls-button-up');
      userEvent.click(upButton);

      expect(testProps.setCurrentPos).toHaveBeenCalled();
  });

  it('handleChangePosition function changes currentPosition when button is clicked', async () => {
    const testProps = {
      ...props,
      currentPos: {
        x: 0,
        y: 0,
      },
      setCurrentPos: jest.fn(),
    };

      render(<Controller {...testProps} />);

      const upButton = await screen.findByTestId('fn-controls-button-up');
      userEvent.click(upButton);

      expect(testProps.setCurrentPos).not.toHaveBeenCalled();
  });

  it('handleBark function has been called when center button is clicked', async () => {
      const { container } = render(<Controller {...props} />);

      const barkButton = await screen.findByTestId('fn-controls-bark');
      userEvent.click(barkButton);

      expect(container).toMatchSnapshot();
  });
});
