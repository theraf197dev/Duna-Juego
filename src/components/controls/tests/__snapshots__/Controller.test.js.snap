// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller tests handleBark function has been called when center button is clicked 1`] = `
<div>
  <div
    class="sc-bcXHqe gzQsTO"
  >
    <button
      class="sc-gswNZR eKcjBW"
      data-testid="fn-controls-button-up"
    >
      <div
        class="sc-dkrFOg eOGSWW"
      />
    </button>
    <button
      class="sc-gswNZR dvMeFY"
      data-testid="fn-controls-button-down"
    >
      <div
        class="sc-dkrFOg iMLtWa"
      />
    </button>
    <button
      class="sc-gswNZR dZEoHc"
      data-testid="fn-controls-button-left"
    >
      <div
        class="sc-dkrFOg etZoki"
      />
    </button>
    <button
      class="sc-gswNZR lkPqEd"
      data-testid="fn-controls-button-right"
    >
      <div
        class="sc-dkrFOg kjDmxe"
      />
    </button>
    <div
      class="sc-eDvSVe dTHyfn"
    >
      <div
        class="sc-hLBbgP kmMqap"
        data-testid="fn-controls-bark"
      />
    </div>
  </div>
</div>
`;

exports[`Controller tests render component 1`] = `
<div>
  <div
    class="sc-bcXHqe gzQsTO"
  >
    <button
      class="sc-gswNZR eKcjBW"
      data-testid="fn-controls-button-up"
    >
      <div
        class="sc-dkrFOg eOGSWW"
      />
    </button>
    <button
      class="sc-gswNZR dvMeFY"
      data-testid="fn-controls-button-down"
    >
      <div
        class="sc-dkrFOg iMLtWa"
      />
    </button>
    <button
      class="sc-gswNZR dZEoHc"
      data-testid="fn-controls-button-left"
    >
      <div
        class="sc-dkrFOg etZoki"
      />
    </button>
    <button
      class="sc-gswNZR lkPqEd"
      data-testid="fn-controls-button-right"
    >
      <div
        class="sc-dkrFOg kjDmxe"
      />
    </button>
    <div
      class="sc-eDvSVe dTHyfn"
    >
      <div
        class="sc-hLBbgP kmMqap"
        data-testid="fn-controls-bark"
      />
    </div>
  </div>
</div>
`;
